src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -187,77 +198,222 @@ def emulate_bundle(self, bundle_fqid: BundleFQID) -> Bundle:
-                SELECT {columns}
-                FROM {self._source.bq_name}.{entity_type}
-                WHERE {uuid_in_list}
-            ''', group_by=pk_column)
-            return rows
+                raise RequirementError('Unexpected link_type', link_type)
+        return entities, inputs, outputs
 
-        with ThreadPoolExecutor(max_workers=config.num_tdr_workers) as executor:
-            futures = {
-                entity_type: executor.submit(retrieve_rows, entity_type, entity_ids)
-                for entity_type, entity_ids in entities.items()
-            }
-        for entity_type, future in futures.items():
-            e = future.exception()
-            if e is None:
-                rows = future.result()
-                entity_ids = entities[entity_type]
-                for i, row in enumerate(rows):
-                    bundle.add_entity(f'{entity_type}_{i}.json', entity_type, row)
-                    entity_ids.remove(row[entity_type + '_id'])
-                reject(bool(entity_ids),
-                       f'Required entities not found in {self._source.bq_name}.{entity_type}: '
-                       f'{entity_ids}')
-            else:
-                log.error('TDR worker failed to retrieve entities of type %r',
-                          entity_type, exc_info=e)
-                raise e
-        return bundle
+    def _dangling_inputs(self, inputs: Entities, outputs: Entities) -> Entities:
+        return {
+            input
+            for input in inputs
+            if input.entity_type.endswith('_file') and input not in outputs
+        }
+
+    def _find_upstream_bundles(self, outputs: Entities) -> Mapping[BundleFQID, Entities]:
+        """
+        Search for bundles containing processes that produce the specified output
+        entities.
+        """
+        output_id = 'JSON_EXTRACT_SCALAR(link_output, "$.output_id")'
+        output_type = 'JSON_EXTRACT_SCALAR(link_output, "$.output_type")'
+        rows = self._run_sql(f'''
+            SELECT links_id, version, {output_id} AS output_id, {output_type} AS output_type
+            FROM {self._source.bq_name}.links AS links
+                JOIN UNNEST(JSON_EXTRACT_ARRAY(links.content, '$.links')) AS content_links
+                    ON JSON_EXTRACT_SCALAR(content_links, '$.link_type') = 'process_link'
+                JOIN UNNEST(JSON_EXTRACT_ARRAY(content_links, '$.outputs')) AS link_output
+                    ON {output_id} IN UNNEST({[output.entity_id for output in outputs]})
+        ''')
+        bundles = defaultdict(set)
+        for row in rows:
+            fqid = BundleFQID(uuid=row['links_id'],
+                              version=row['version'].strftime(self.timestamp_format))
+            ref = EntityReference(entity_id=row['output_id'],
+                                  entity_type=row['output_type'])
+            bundles[fqid].add(ref)
+        missing = outputs - set.union(*bundles.values())
```

Comment:
(my approach also caused a bug when no bundles were found, since Python doesn't seem to understand that the union of a collection of empty sets is well-defined, namely, the empty set)